0000-richer-documents.rst

```diff
@@ -0,0 +1,300 @@
+to emit error messages with bits of expressions, types and more
+embedded in the documents, as AST values, as opposed to their
+textual form which is not amenable to any form of inspection or
+post-processing.
+
+Motivation
+----------
+
+GHC's whole pretty printing infrastructure is based around the following
+types:::
+
+    data Doc
+      = Empty                                            -- empty
+      | NilAbove Doc                                     -- text "" $$ x
+      | TextBeside !TextDetails {-# UNPACK #-} !Int Doc  -- text s <> x
+      | Nest {-# UNPACK #-} !Int Doc                     -- nest k x
+      | Union Doc Doc                                    -- ul `union` ur
+      | NoDoc                                            -- The empty set of documents
+      | Beside Doc Bool Doc                              -- True <=> space between
+      | Above Doc Bool Doc                               -- True <=> never overlap
+
+    data SDocContext = SDC { ... }
+    newtype SDoc = SDoc { runSDoc :: SDocContext -> Doc }
+
+The exact contents of ``SDocContext`` is not very important, one just needs
+to know that it gives access to all sorts of pretty-printing and GHC
+related configuration. These two types allow us to specify how sub-documents
+should be placed relative to one another once rendered and have been great to
+describe all sorts of textual documents in GHC.
+
+On the other hand, authors of tools in the IDE/editor-integration space
+have been wondering about the possibility of enriching those error
+documents, with several applications in mind -- we will mention a few in
+the **Effects and Interactions** section. All of those applications require
+that we keep some information around, stored in good old ADTs, that we can
+freely inspect and consume: bits of expressions, types, constraints, etc.
+
+To bridge that gap, this proposal suggests that we add support for
+embedding annotations (non-textual values) directly into our documents,
+as described in the next section.
+
+Proposed Change Specification
+-----------------------------
+
+This proposal suggests that we modify ``Doc`` to extend it with a
+constructor for embedding arbitrary values into documents, which we refer
+to as "annotations".::
+
+    data Doc a
+      = Empty
+        -- empty
+      | ...
+      | Pure a -- new! embed an annotation of type 'a'
+
+Such a document model comes with useful instances for standard
+classes, among which ``Functor``, ``Aplicative``, ``Monad``, ``Foldable``
+and ``Traversable``, giving us standard names for functions that
+transform annotations into other annotations or into whole new sub-documents.
+
+``SDoc`` would similarly be updated to take a type parameter for the
+annotation type:::
+
+    newtype SDoc a = SDoc { runSDoc :: SDocContext -> Doc a }
+
+and would come with similar instances as ``Doc`` (since with this definition,
+``SDoc ~ ReaderT SDocContext Doc``), except for ``Foldable`` and
```

Comment:
I think you mean `Coercible` here, not `~`.

Reply:
Yes, it was a "handwavy `~`", which indeed corresponds to `Coercible`.